internal/utils.go

```diff
@@ -0,0 +1,15 @@
+package internal
+
+import (
+	"github.com/multiformats/go-multihash"
+)
+
+func Sha256Multihash(mh multihash.Multihash) multihash.Multihash {
+	prefix := []byte("CR_DOUBLEHASH")
+	mh, err := multihash.Sum(append(prefix, mh...), multihash.DBL_SHA2_256, 30)
```

Comment:
is there a reason we're saying this has a length of 30?

Reply:
yeah, since the multihash automatically prepends the `code` and `length` of the data, the returned multihash is of length `32`. If I set the length to `32` (which I initially tried) it causes the returned multihash to be of length 34, which then causes a panic in `routingTable.NearestPeers` when it calls `CommonPrefixLen` to compare the multihash with the local peer ID, since the lengths are mismatched.

the other workaround I thought off is slicing off the first 2 bytes of the multihash / only using the digest, but I'm not sure if that would have unintended consequences. any thoughts on what method would be better?

Reply:
```go
func Encode(buf []byte, code uint64) ([]byte, error) {
	// FUTURE: this function always causes heap allocs... but when used, this value is almost always going to be appended to another buffer (either as part of CID creation, or etc) -- should this whole function be rethought and alternatives offered?
	newBuf := make([]byte, varint.UvarintSize(code)+varint.UvarintSize(uint64(len(buf)))+len(buf))
	n := varint.PutUvarint(newBuf, code)
	n += varint.PutUvarint(newBuf[n:], uint64(len(buf)))

	copy(newBuf[n:], buf)
	return newBuf, nil
}
```

The [`Encode()`](https://github.com/multiformats/go-multihash/blob/196bbeb81a568a71ec224866ccc68c1f28a272d8/multihash.go#L248) function basically takes the digest and append it to the code, hence returning something of the format: [code, digest]. The [code](https://github.com/multiformats/go-multihash/blob/196bbeb81a568a71ec224866ccc68c1f28a272d8/multihash.go#L67) is represented by 2 bytes.

We don't want the digest to start with a constant, as this digest determine the location of the provider records in the DHT. Having a common prefix (code) to all digest would mean that all Double Hashed Provider Records are allocated to a restricted location of the keyspace. We don't need to have the code in the output. We could update the code as follow:

```go
mh, err := multihash.Sum(append(prefix, mh...), multihash.DBL_SHA2_256, KEYSIZE)
	if err != nil {
		// this shouldn't ever happen
		panic(err)
	}
        // don't return the code prefix, only return the last KEYSIZE bytes
	return mh[len(mh)-KEYSIZE:]
```
`KEYSIZE` should be a constant value defining the DHT key size. Ideally this constant already exists somewhere in the DHT implementation or it is passed down by kubo. Its value should be 32 for now, but if the DHT keysize was to change this piece of code shouldn't need to be updated, updating the constant should be enough.

Reply:
@guillaumemichel that works, but the issue is now the function returns a `multihash.Multihash` that isn't actually a multihash (since the code is gone). the purpose from what I remember of using `multihash.Sum` was that we could specify the hash was of type `DBL_SHA2_256`, whereas previously it was just using sha256 (which is effectively the same as doing `multihash.Sum` then choosing the last 32 bytes).  should I just change the function back to do a `sha256` sum?

Reply:
Yes, it should be fine to change the function back to `sha256`